### ISSUE TITLE
Cluster support - machine failures & preemptible setups

### ISSUE THREAD

**Issue by chrisroat** [datajoint/datajoint-python#665]
When running in a cloud environment:
- machines often fail, and the cluster will provision new ones
- it can save a lot of money to use preemptible machines, which means a machine may go away at any time

The `reserve_jobs` feature of `AutoPopulate` datajoint tables could be expanded to operate more efficiently in these contexts.  Since work should be idempotent (any random seeds for bootstrapping, etc should be deterministic), it is technically allowed to have multiple machines work on the same processing step.

One idea would be to have the job reservation have an expiration, and workers can take reserved work that is past its expiration date.

**Comment by dimitri-yatsenko** [comment]
Thank you, @chrisroat Yes many extension can be developed for cloud computing and cluster computing support and greater efficiency.

The job reservation tables already contain the timestamp for when the job was started and the database connection id, hostname, and process id. This information allows checking if the process doing  the job is still alive.  We periodically clear any reservation whose database connection has gone away, for example.

**Comment by chrisroat** [comment]
I completely missed the cleanup, as that should cover the basic case -- where is that happening?

**Comment by ixcat** [comment]
I believe @dimitri-yatsenko is referring to a secondary process used in a real-world pipeline, but not a built-in feature of DataJoint.

**Comment by dimitri-yatsenko** [comment]
Yes. Normal execution requires no cleanup. If a job completes normally, it removes its reservation record. If the job ends in an error, it replaces the status of the reservation to `error` and so the job becomes unavailable. Also, if the job simply crashes, then its record remains with the status `reserved`. But you can know which ones have crashed by checking  the database connection numbers in `schema.jobs` and using `dj.kill()` to view which ones are no longer connected.

**Comment by dimitri-yatsenko** [comment]
We just occasionally do `schema.jobs.delete()` to delete all reservations when no jobs are being performed.

**Comment by ixcat** [comment]
treating as 'main' issue for interrelated issues: #665 #550 #510 and to lesser extent #749

**Comment by yambottle** [comment]
This issue is closed because there has been no activity for more than 365 days, please reopen if it's still applicable.
